File: python/main_torchscript.py
import os
import sys

import py3nvml.py3nvml as nvml
import torch
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from torch import nn
from torch.cuda.nvtx import range_push, range_pop

from cuda_profiling import GPUTimer, get_used_cuda_mem, get_gpu_utilization, cuda_profiler_start, cuda_profiler_stop
from pascal import VOCDetection
from resnet import ResNet50

DEVICE = int(os.environ["DEVICE"])
print(f"device {DEVICE}")
gpu_timer = GPUTimer(DEVICE)

nvml.nvmlInit()
nvml_handle = nvml.nvmlDeviceGetHandleByIndex(DEVICE)


def train(
    model,
    trainloader,
    testloader,
    optimizer,
    criterion,
    epochs,
    batch_size,
    monitoring_step,
    output_file=None,
):
    cuda_profiler_start()
    for epoch in range(epochs):
        elapsed_time = 0
        model.eval()
        sample_count = tp_count = loss_val = 0
        used_mem = 0

        with torch.no_grad():
            for batch_n, (inputs, labels) in enumerate(testloader):
                range_push(f"train epoch {epoch} batch {batch_n}")
                range_push(f"batch load")
                gpu_timer.start()

                inputs = inputs.to(f"cuda:{DEVICE}")
                labels = labels.to(f"cuda:{DEVICE}")

                range_pop()

                outputs = model(inputs)
                loss = criterion(outputs, labels)
                _, predicted = torch.max(outputs.data, 1)

                range_pop()

                gpu_timer.stop()

                loss_val += loss.item()
                tp_count += (predicted == labels).sum().item()
                sample_count += batch_size
                elapsed_time += gpu_timer.elapsed_time()
                used_mem += get_used_cuda_mem(DEVICE)

        print(
            f"[EVAL] "
            f"avg loss: {loss_val / float(sample_count):.10f}, "
            f"accuracy: {100.0 * float(tp_count) / sample_count:.6f}%, "
            f"avg sample time: {elapsed_time / sample_count:.6f}ms, "
            f"avg used mem: {used_mem / (sample_count / batch_size)}mb, "
            f"avg gpu util: {get_gpu_utilization(nvml_handle)}%",
            # file=output_file,
            # flush=True,
        )
    cuda_profiler_stop()


object_categories = [
    "aeroplane",
    "bicycle",
    "bird",
    "boat",
    "bottle",
    "bus",
    "car",
    "cat",
    "chair",
    "cow",
    "diningtable",
    "dog",
    "horse",
    "motorbike",
    "person",
    "pottedplant",
    "sheep",
    "sofa",
    "train",
    "tvmonitor",
]

object_categories_idx = dict(
    list(zip(object_categories, range(len(object_categories))))
)


def transform_pascal(x):
    name = x["annotation"]["object"][0]["name"]
    idx = object_categories_idx[name]
    return idx


def main():
    dataset_name, run_n, epochs, resolution, batch_size = (
        sys.argv[1],
        sys.argv[2],
        int(sys.argv[3]),
        int(sys.argv[4]),
        int(sys.argv[5]),
    )
    # lost .5 ms with smaller batch size for mnist
    monitoring_step = 20

    transform = [
        transforms.ToTensor(),
        # transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
    ]

    if dataset_name == "cifar10":
        trainset = torchvision.datasets.CIFAR10(
            root="../data",
            train=True,
            download=True,
            transform=transforms.Compose(transform),
        )
        testset = torchvision.datasets.CIFAR10(
            root="../data",
            train=False,
            download=True,
            transform=transforms.Compose(transform),
        )
        in_channels = 3
        model = ResNet50(in_channels=in_channels, num_classes=10)
    elif dataset_name == "stl10":
        trainset = torchvision.datasets.STL10(
            root="../data",
            split="train",
            download=True,
            transform=transforms.Compose(transform),
        )
        testset = torchvision.datasets.STL10(
            root="../data",
            split="test",
            download=True,
            transform=transforms.Compose(transform),
        )
        in_channels = 3
        model = ResNet50(in_channels=in_channels, num_classes=10)
    elif dataset_name == "mnist":
        trainset = torchvision.datasets.MNIST(
            root="../data",
            train=True,
            download=True,
            transform=transforms.Compose(transform),
        )
        testset = torchvision.datasets.MNIST(
            root="../data",
            train=False,
            download=True,
            transform=transforms.Compose(transform),
        )
        in_channels = 1
        model = ResNet50(in_channels=in_channels, num_classes=10)
    elif dataset_name == "pascal":
        transform.insert(0, transforms.Resize((resolution, resolution)))
        trainset = VOCDetection(
            root="../data/",
            year="2012",
            image_set="train",
            download=True,
            transform=transforms.Compose(transform),
            target_transform=transform_pascal,
        )
        testset = VOCDetection(
            root="../data/",
            year="2012",
            image_set="val",
            download=False,
            transform=transforms.Compose(transform),
            target_transform=transform_pascal,
        )
        in_channels = 3
        model = ResNet50(in_channels=in_channels, num_classes=20)
    else:
        raise Exception("unsupported dataset")

    trainloader = torch.utils.data.DataLoader(
        trainset, batch_size=batch_size, shuffle=True, num_workers=1
    )
    testloader = torch.utils.data.DataLoader(
        testset, batch_size=batch_size, shuffle=False, num_workers=1
    )

    # traced_script_module = torch.jit.script(model)
    inputs = (torch.rand(batch_size, 3, resolution, resolution),)
    check_inputs = [
        (torch.rand(batch_size // 2, 3, resolution, resolution),),
        (torch.rand(batch_size // 4, 3, resolution, resolution),),
    ]

    traced_script_module = torch.jit.trace(model, inputs, check_inputs=check_inputs)
    traced_script_module.save(f"traced_resnet50_{resolution}.pt")

    traced_script_module = traced_script_module.to(f"cuda:{DEVICE}")
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
    train(
        traced_script_module,
        trainloader,
        testloader,
        optimizer,
        criterion,
        epochs,
        batch_size,
        monitoring_step,
        open(
            f"profiles/resolution/run_pytorch_{dataset_name}_{batch_size}_{resolution}.csv",
            "w",
        ),
    )


if __name__ == "__main__":
    main()

File: cpp/lib_torch/src/main_torchscript.cpp
#include "cifar10.h"
#include "cuda_helper.h"
#include "cuda_profiling.h"
#include "nvml.h"
#include "pascal.h"
#include "transform.h"

#include <cuda_profiler_api.h>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <nvtx3/nvToolsExt.h>
#include <torch/script.h>
#include <torch/torch.h>
#include <tuple>

using namespace torch;
using torch::data::datasets::MNIST;
using transform::RandomHorizontalFlip;

template <typename DatasetType>
using TrainDataset = data::datasets::MapDataset<
    data::datasets::MapDataset<DatasetType, RandomHorizontalFlip>,
    data::transforms::Stack<data::Example<>>>;
template <typename DatasetType>
using TestDataset =
    data::datasets::MapDataset<DatasetType, data::transforms::Stack<data::Example<>>>;

template <typename DatasetType>
void train(
    torch::jit::Module *model,
    TestDataset<DatasetType> train_dataset,
    TestDataset<DatasetType> test_dataset,
    int64_t batch_size,
    int monitoring_step,
    double learning_rate,
    size_t num_epochs,
    torch::Device device,
    std::ostream &output_file) {

    auto num_test_samples = test_dataset.size().value();
    auto test_loader = torch::data::make_data_loader<torch::data::samplers::SequentialSampler>(
        std::move(test_dataset), batch_size);

    std::string nvtx_message;
    auto gpu_timer = GPUTimer();
    cudaProfilerStart();

    double loss_val, accuracy, running_loss;
    int tp_count, running_tp_count, running_sample_count, sample_count;
    double total_time;
    double elapsed_time;
    double used_mem = 0, running_used_mem = 0;
    int batch_n = 0;
    std::vector<torch::jit::IValue> inputs;

    for (size_t epoch = 0; epoch != num_epochs; ++epoch) {
        model->eval();
        torch::NoGradGuard no_grad;

        total_time = sample_count = tp_count = loss_val = 0;
        used_mem = 0;

        for (const auto &batch : *test_loader) {
            nvtx_message = std::string(
                "eval epoch " + std::to_string(epoch) + " batch " +
                std::to_string(batch_n));
            nvtxRangePushA(nvtx_message.c_str());
            nvtxRangePushA("batch load");

            gpu_timer.start();

            auto data = batch.data.to(device);
            auto target = batch.target.to(device);
            nvtxRangePop();

            inputs.clear();
            inputs.push_back(data);
            auto output = model->forward(inputs).toTensor();
            auto loss = torch::nn::functional::cross_entropy(output, target);
            auto prediction = output.argmax(1);
            tp_count += prediction.eq(target).sum().template item<int64_t>();

            nvtxRangePop();

            gpu_timer.stop();

            loss_val += loss.template item<double>();
            sample_count += batch_size;
            total_time += gpu_timer.elapsed();
            used_mem += get_used_cuda_mem();
        }

        output_file << "[EVAL] avg loss: " << std::setw(4) << loss_val / sample_count
                    << ", accuracy: " << 100.f * tp_count / sample_count << "%"
                    << ", avg sample time: " << total_time / sample_count << "ms"
                    << std::defaultfloat
                    << ", avg used mem: " << used_mem / (sample_count / batch_size) << "mb"
                    << ", avg gpu util: " << get_gpu_utilization() << "%" << std::endl;
        output_file.flush();

    }
}

int main(int argc, char *argv[]) {
    auto cuda_available = torch::cuda::is_available();
    torch::Device device(cuda_available ? torch::kCUDA : torch::kCPU);
    std::cout << (cuda_available ? "CUDA available. Training on GPU." : "Training on CPU.") << '\n';

    int64_t batch_size = std::stoi(std::getenv("BATCH_SIZE"));
    int num_classes;
    int num_epochs = std::stoi(std::getenv("EPOCHS"));
    int monitoring_step = 20;

    double learning_rate = 1.0 / std::stoi(std::getenv("INV_LEARNING_RATE"));

    torch::jit::script::Module torch_script_model;
    try {
        std::stringstream ss;
        ss << "traced_resnet50_" << std::getenv("RESOLUTION") << ".pt";
        // Deserialize the ScriptModule from a file using torch::jit::load().
        torch_script_model = torch::jit::load(ss.str());
    } catch (const c10::Error &e) {
        std::cerr << "error loading the model\n";
        exit(EXIT_FAILURE);
    }

    std::stringstream ss;
    ss << "profiles/run_torchscript"
       << "_" << batch_size << "_" << std::getenv("RESOLUTION") << ".csv";
    std::ofstream output_file(ss.str());

    if (strcmp(argv[1], "pascal") == 0) {
        auto model = &torch_script_model;
        model->to(device);

        std::cout << "== PASCAL training with TorchScript ==" << std::endl;
        TestDataset<PASCAL> train_dataset = PASCAL("../data/VOCdevkit/VOC2012", PASCAL::kTrain)
                                                .map(torch::data::transforms::Stack<>());
        TestDataset<PASCAL> test_dataset = PASCAL("../data/VOCdevkit/VOC2012", PASCAL::kVal)
                                               .map(torch::data::transforms::Stack<>());

        train<PASCAL>(
            model,
            train_dataset,
            test_dataset,
            batch_size,
            monitoring_step,
            learning_rate,
            num_epochs,
            device,
            output_file);
    } else {
        exit(EXIT_FAILURE);
    }
}
